### ISSUE TITLE
[R-Forge #4903] set encoding in fread

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#563]
Submitted by: michele de meo; Assigned to: Nobody; [R-Forge link](https://r-forge.r-project.org/tracker/index.php?func=detail&aid=4903&group_id=240&atid=978)

It's useful to specify the encoding before importing   a csv.
For example: 

fread(... , encoding='UTF8').

In this way we can avoid the boring use of file function in read.table (not supported in fread ):

read.table(file("mycsv.csm", encoding='UTF8'), ... ).

Michele De Meo

**Comment by sirvyd** [comment]
+1

**Comment by harmonica2** [comment]
+1

**Comment by zeltak** [comment]
+1

**Comment by MarcinKosinski** [comment]
+1

**Comment by jomuller** [comment]
+1 Would be really usefull. I started to hack locally the code but it's too hard for me because it call a C function (readfile) and I don't know how (for the moment) how to handle encoding in C. If any one some knowledge in C...

**Comment by grgurev** [comment]
+1

**Comment by lucarno** [comment]
+1

**Comment by seyedamo** [comment]
+1

**Comment by jcizel** [comment]
+1

**Comment by Alectoria** [comment]
+1

**Comment by tophcito** [comment]
+1

**Comment by clarkdk** [comment]
+1

**Comment by JohnsonHsieh** [comment]
+1

**Comment by whizzalan** [comment]
+1

**Comment by ysgit** [comment]
+1

**Comment by ZeroStack** [comment]
+1

**Comment by rshmyrev** [comment]
+1

**Comment by dbuijs** [comment]
+1, and in the meantime, a workaround:

You need iconv available from the console. You can check this with the following command:

    Sys.which("iconv")

As long as this gives you a path to a binary, the following will work:

    # From ISO-8859-1, To UTF-8
    new.data.table <- fread("iconv -f ISO-8859-1 -t UTF-8 mytextfile.txt")

**Comment by shrektan** [comment]
+ 1

**Comment by panda2727** [comment]
+1

**Comment by duf59** [comment]
+1

**Comment by romunov** [comment]
+1

note however that in `?Encoding`, `UTF-8` is used (not `UTF8`).

**Comment by rentrop** [comment]
+1

**Comment by leoluyi** [comment]
+1

**Comment by arunsrinivasan** [comment]
Tl;dr 

Could you please test your files with `encoding = "UTF-8"` or `"Latin-1"` and write back here as to whether it solves the issue (especially on windows)? Thanks.

---

About the fix: 

Looking at `read.table()` function, the encoding is set upfront with `Encoding(..) <- ...`. And looking at the source of `Encoding<-`, it calls an internal function `do_setencoding`. Looking at the source of that function, https://github.com/wch/r-source/blob/ca5348f0b5e3f3c2b24851d7aff02de5217465eb/src/main/util.c#L1115, it seems quite straightforward to fix this (I think).. through the use of `mkCharLenCE` function that R's C-API seems to expose.

`fread()` gains a new argument `encoding` with valid values `unknown`, `UTF-8` and `Latin-1`. The default behaviour hasn't changed. Windows user will have to set the encoding argument explicitly. This could change in the future (perhaps also testing for performance hit would be a deciding factor).

**Comment by clarkdk** [comment]
Arun, There's a typo in README.md

27. fread() gains eocnding argument. ...  eocnding --> encoding

**Comment by jangorecki** [comment]
if speaking about the typos there is one missing `()` next to `flush.console` in `bmerge` function: https://github.com/Rdatatable/data.table/blob/7f78b1b019ae04593c7526d244a9a8559e1be471/R/bmerge.R#L102

**Comment by dbuijs** [comment]
Thank you for this!!!!

A question of clarification: does the `encoding` option in `fread` just set the marked encoding, or does it convert from what you've declared to the system's native encoding?

**Comment by andreasio** [comment]
I have an iso-8850-1 windows file, that I can't fread correctly on linux. This workaround (https://github.com/Rdatatable/data.table/issues/563#issuecomment-112423962) works. And read.csv works (i.e. produces the correct øæå letters, instead of e.g. \xe6), but fread(... encoding = "Latin-1") does not.
